### ISSUE TITLE
VS Code hang

### ISSUE THREAD

**Issue by gregvanl** [Microsoft/vscode#62970]
- VSCode Version: 1.29.0 Insiders
- OS Version: Windows

Several hangs in a row. Attached are the log files from two of the hangs which contain the error `[error] [VS Code]: detected unresponsive`.

[InsidersHangs.zip](https://github.com/Microsoft/vscode/files/2573731/InsidersHangs.zip)

**Comment by sgyzetrov** [comment]
Hi, I am also experiencing hang. VS Code version Version 1.29.0, macOS 10.13.6 (17G3025)

But unfortunately, because the window was unresponsive, I cannot get developer tool to launch, so I did not get any CPU profiling. But I do have control over other VS Code windows running separately to the window that frozen. Will profiling other windows produce useful information? Also, during the hang, Code Helper spiked to 100%.

**Comment by joseph1125** [comment]
1.29 stable with Windows 10 also experience this problem

**Comment by tomaxsas** [comment]
same here with Win 10 and 1.29.0

**Comment by jrieken** [comment]
@sgyzetrov @joseph1125 @tomaxsas Only CPU profiling will help us understand this. If the hang is very bad, then restart VS Code, start dev tools and keep them in the background, and profile when the editor is unresponsive.

**Comment by sgyzetrov** [comment]
> @sgyzetrov @joseph1125 @tomaxsas Only CPU profiling will help us understand this. If the hang is very bad, then restart VS Code, start dev tools and keep them in the background, and profile when the editor is unresponsive.

Yeah...I did that, just now. After letting dev tool run in the background for more than 3 hrs, finally VS Code hang again (I was not even working on it, I was browsing Reddit on another macOS desktop), and this time, CPU profiler **got stuck** too. 

I clicked start (it then turned into a red 'stop'), I then let it run for 57 sec, nothing happened, so I clicked 'stop'. BUT then it just STUCK. There is no info coming up, I tried to start a new profile, but the button is a grey 'start'. And `collect garage`, `clear all profiles` were all unresponsive. `start CPU profiling` told me another profiling is already running, but I definitely clicked stop.

<img width="500" alt="activity_monitor__all_processes_-2" src="https://user-images.githubusercontent.com/34640565/48421577-beaeba00-e797-11e8-997f-1770e00092f0.png">
<img width="495" alt="activity_monitor__all_processes_" src="https://user-images.githubusercontent.com/34640565/48421578-bf475080-e797-11e8-845f-c3029236c719.png">
 Looks to me there is a serious memory leak.
I will wait for another hang and keep trying to acquire the CPU profile you want, stand by.

**Comment by sgyzetrov** [comment]
Had to force quit Code Helper and VS Code to let my fan slow down...

**Comment by jrieken** [comment]
> I then let it run for 57 sec,

Try to profile for much shorter then, 5-10 seconds and even then it can take a very long time to finally save the profile.

**Comment by VishwaRajGit** [comment]
I have also experienced this twice just now.
VS Code freezing when it's not even being worked on.
1.29 on Windows 10

**Comment by tecosaur** [comment]
I'm getting this severely too. Is there any way I can roll back to the previous version?

**Comment by jrieken** [comment]
@VishwaRajGit @tecosaur please help us profile this. 

* start dev tools via `F1 > Developer Tools` 
* open the cpu profiler via 'More Tools > CPU Profiler' in the overflow menu (see https://github.com/Microsoft/vscode/issues/62970#issuecomment-438164153)
* pop-out and minimise the dev tools window
* use vs code as usual and wait for it to slow down
* when it hangs, freezes, or slows down, then go to dev tools and hit 'Start' in the CPU Profiler tab
* profile for 5-10 seconds and press the red botton to stop profiling
* be patient and when the profile appears, select 'Save', and share the profile here

![nov-13-2018 18-08-19](https://user-images.githubusercontent.com/1794099/48430246-27ce0780-e76f-11e8-843f-10cd9279b87f.gif)


Thanks for your help.

**Comment by jrieken** [comment]
everyone also attach the list of extensions you are using, simply run `code --list-extensions`.

**Comment by dgolovin** [comment]
@jrieken After vs code window was frozen, I was following the steps above and profiler just got into state shown below. I am still waiting for 'Save' to appear, but it seems never going to happen.
![image](https://user-images.githubusercontent.com/620330/48436378-6645d100-e733-11e8-9962-5ba68333a917.png)

**Comment by gitawego** [comment]
my vscode got hangs as well with new version.
my extensions: 
dbaeumer.vscode-eslint
donjayamanne.githistory
eamodio.gitlens
EditorConfig.EditorConfig
eg2.tslint
esbenp.prettier-vscode
formulahendry.auto-close-tag
formulahendry.auto-rename-tag
mikegleasonjr.theme-go
ms-python.python
ms-vscode.Go
msjsdiag.debugger-for-chrome
PeterJausovec.vscode-docker
redhat.java
robertohuertasm.vscode-icons
rust-lang.rust
streetsidesoftware.code-spell-checker

**Comment by khayuenkam** [comment]
Freeze quite frequently to me as well. 
I have attached the cpu profile. 
[profile.txt](https://github.com/Microsoft/vscode/files/2578353/CPU-20181113T230227.txt)

VSCode Version: 1.29.0
OS Version: MacOS 10.13.6

Code Extensions:
- CoenraadS.bracket-pair-colorizer
- eamodio.gitlens
- jcanero.hoogle-vscode
- justusadam.language-haskell
- sdras.night-owl
- Vans.haskero
- vscodevim.vim

**Comment by sgyzetrov** [comment]
Hi folks, 

It sees VS Code will reproduce this hang whenever it is not being worked on. Although strangely if your workspace only contains `markdown` or `HTML` this issue would not occur. Since I did not experience hang in separate VS Code windows where I was editing `.md` or `.html` files.

The window I got hang was the one I was editing `.tex` files with `latex-workshop` plugin.

@jrieken Johannes, I finally got the CPU profile. This time I profiled for 5 sec, after waiting for about 5 min I was able saved it. 

My extensions are listed below.

```
Gimly81.matlab
austin.code-gnu-global
DavidAnson.vscode-markdownlint
DotJoshJohnson.xml
gnhuy91.theme-oceanicnext-sublime
James-Yu.latex-workshop
magicstack.MagicPython
mitaki28.vscode-clang
ms-python.python
ms-vscode.azure-account
ms-vscode.cpptools
naumovs.theme-oceanicnext
PeterJausovec.vscode-docker
PKief.material-icon-theme
redhat.java
tht13.python
vscjava.vscode-java-debug
vscjava.vscode-java-pack
vscjava.vscode-java-test
vscjava.vscode-maven
yzhang.markdown-all-in-one
```

CPU profile with `.txt` extension added
[CPU-20181114T084129.cpuprofile.txt](https://github.com/Microsoft/vscode/files/2578720/CPU-20181114T084129.cpuprofile.txt)

**Comment by alex-ketch** [comment]
Similar to @sgyzetrov I've noticed VS Code reliably freezing if I switch focus to a different application and then come back to it.

Attached is a CPU profile from the latest freeze. I only tabbed over to Safari for a few seconds.
Please let me know if I can help any further.

VSCode Version: 1.30.0-insider
OS Version: MacOS 10.14.1

<details>
<summary>Installed Extensions</summary>

- `Zarel.sublime-commands`
- `Zignd.html-css-class-completion`
- `aaron-bond.better-comments`
- `abusaidm.html-snippets`
- `alanz.vscode-hie-server`
- `alefragnani.Bookmarks`
- `alefragnani.project-manager`
- `alqiu.ts-loader-problem-matcher`
- `andys8.jest-snippets`
- `arcticicestudio.nord-visual-studio-code`
- `bierner.markdown-preview-github-styles`
- `christian-kohler.npm-intellisense`
- `christian-kohler.path-intellisense`
- `chrmarti.regex`
- `daltonjorge.scala`
- `dbaeumer.vscode-eslint`
- `deerawan.vscode-dash`
- `dinhani.divider`
- `donjayamanne.githistory`
- `eamodio.find-related`
- `eamodio.gitlens`
- `EditorConfig.EditorConfig`
- `eg2.tslint`
- `esbenp.prettier-vscode`
- `felipe.nasc-touchbar`
- `formulahendry.auto-rename-tag`
- `gamunu.vscode-yarn`
- `GitHub.vscode-pull-request-github`
- `HookyQR.beautify`
- `infeng.vscode-react-typescript`
- `jcanero.hoogle-vscode`
- `joelday.docthis`
- `jpoissonnier.vscode-styled-components`
- `jtlowe.vscode-icon-theme`
- `justusadam.language-haskell`
- `kevinkyang.auto-comment-blocks`
- `kumar-harsh.graphql-for-vscode`
- `maptz.camelcasenavigation`
- `mgmcdermott.vscode-language-babel`
- `mrmlnc.vscode-stylefmt`
- `ms-vsliveshare.vsliveshare`
- `msjsdiag.debugger-for-chrome`
- `naumovs.color-highlight`
- `nmsmith89.incrementor`
- `nwolverson.ide-purescript`
- `nwolverson.language-purescript`
- `Orta.vscode-jest`
- `PeterJausovec.vscode-docker`
- `quicktype.quicktype`
- `ryanluker.vscode-coverage-gutters`
- `Shan.code-settings-sync`
- `shaunlebron.vscode-parinfer`
- `shinnn.stylelint`
- `silvenon.mdx`
- `sleistner.vscode-fileutils`
- `stringham.move-ts`
- `stubailo.ignore-gitignore`
- `teabyii.ayu`
- `vscodevim.vim`
- `WallabyJs.quokka-vscode`
- `wix.vscode-import-cost`
- `wmaurer.change-case`
- `xabikos.JavaScriptSnippets`
- `xabikos.ReactSnippets`
- `Yummygum.city-lights-icon-vsc`
- `yzhang.markdown-all-in-one`
- `zhuangtongfa.Material-theme`

</details>
<br />

CPU profile with .txt extension added
[CPU-20181113T205652.cpuprofile.txt](https://github.com/Microsoft/vscode/files/2578905/CPU-20181113T205652.cpuprofile.txt)

**Comment by roblourens** [comment]
@alex-ketch and @khayuenkam's traces appear to show the same as others I've seen, basically just tons of IPC message receiving. I see NotificationsToasts.purgeNotification showing up a lot? I think that's `workbench.main.js:3620`

This commit https://github.com/Microsoft/vscode/commit/9efef6a484cda14209399356ad814a4f0248b0f6 looks a bit suspicious considering it involves the window onDidChangeFocus, but I have no clue what this would have to do with IPC traffic. 

Did anyone have notifications visible when this happened?

@mjbvz did you get anywhere when looking at the window switching issue?

**Comment by MeikTranel** [comment]
Don't know if this helps, but immediately after hanging i took a peek in Resource Monitor (My disk IO went to 100% in taskmgr) and found this.

![image](https://user-images.githubusercontent.com/6466560/48468636-b9c82580-e7ec-11e8-9625-4f1b1530f662.png)

**Comment by sgyzetrov** [comment]
Hi all,

I have noticed there is little useful information in my previous profile, possibly because the profile time is too short. I am uploading a much larger profile (over 30 sec). @jrieken @roblourens I hope it will be of help. 

[CPU-20181114T160347.cpuprofile.txt](https://github.com/Microsoft/vscode/files/2579716/CPU-20181114T160347.cpuprofile.txt)

**Comment by edvinv** [comment]
Same issue here on ubuntu 18.04. Not sure, but I think that it hangs when I work on other application for some time and then switch back to vscode.

**Comment by jrieken** [comment]
Thanks @sgyzetrov. That profile shows much more. Jus a QQ is that 1.29-stable or 1.30-insiders? 

Yours and other traces hint towards a storm of IPC messages that overwhelm the renderer process. Can please run VS Code with extension host communication logging enabled, like so: 

`code --logExtensionHostCommunication`

That will print all messages that are exchanged between the renderer and extension process to the dev  tool console. Again, open dev tools early and when the freeze happened, then select the console tab, right click in there, and select save as and share the log. Thanks for helping us.

**Comment by jrieken** [comment]
Thanks @sgyzetrov. Please try `code --logExtensionHostCommunication`

**Comment by sgyzetrov** [comment]
Will do, standing by for the next freeze.

**Comment by jrieken** [comment]
@sgyzetrov Sorry, from trace. This isn't between the extension host and the render but between the main process and the renderer. So, logging the extension host communication won't add more info. 

Behaviour does look like a listener leak. Yesterday, I have pushed https://github.com/Microsoft/vscode/commit/b548ced43cb1c9f86f3a013e20ccd8613fbd4138 which is a check for leaking listeners. @sgyzetrov and *everyone else* please use todays Insiders build and keep an eye on the dev tools console. It will print warnings when listeners appear to be leaking: http://code.visualstudio.com/insiders/

**Comment by neildom** [comment]
@roblourens Yes I had three notifications open. One relating to a Jest extension and another two relating to a Flow extension. I moved to a new PC recently and I'm using wsl so I expected that to be the cause of the notifications. In the previous version of vscode, I dismissed the notifications but I read in the release notes for this version that even notifications with buttons should disappear so I have been ignoring the notifications.

**Comment by jrieken** [comment]
Thanks @neilsustc. Also, please try http://code.visualstudio.com/insiders/ keep an eye open for leakage warnings.

**Comment by jrieken** [comment]
Not yet sure what happened but the attached movie shows how tabbing between two windows increases the number of listeners drastically (from 1789 to 5689 with not many switches). [cmd-tab-party.mov.zip](https://github.com/Microsoft/vscode/files/2580584/cmd-tab-party.mov.zip)

As @roblourens already pointed out, this code is fishy https://github.com/Microsoft/vscode/blob/c9408a694059aed5216f188141b829b877755cae/src/vs/workbench/browser/parts/notifications/notificationsToasts.ts#L214-L228

I don't know yet what it is but whenever this [if-condition](https://github.com/Microsoft/vscode/blob/c9408a694059aed5216f188141b829b877755cae/src/vs/workbench/browser/parts/notifications/notificationsToasts.ts#L222) is true and focus has been gained then listeners recursively add them as part of handling the event. Somehow auto hiding notification toats aren't disposed (or disposed too late) and that causes this crazy listener accumulation.

**Comment by jrieken** [comment]
![nov-14-2018 15-29-21](https://user-images.githubusercontent.com/1794099/48488918-1bef4d80-e822-11e8-8107-6ad437914602.gif)

I can reliably reproduce this. In above gif I have added a message that prints whenever a new listener is added, I have also tweaked the purge timeout to make things happen faster. 

1. have a sticky notification, like the update notifications or one with buttons
2. when it pops up and within the purge timeout (15sec) focus another app
3. now we have a sticky notification but no window focus ([L222](https://github.com/Microsoft/vscode/blob/c9408a694059aed5216f188141b829b877755cae/src/vs/workbench/browser/parts/notifications/notificationsToasts.ts#L222)) which means we add another listener ([L223](https://github.com/Microsoft/vscode/blob/c9408a694059aed5216f188141b829b877755cae/src/vs/workbench/browser/parts/notifications/notificationsToasts.ts#L223)) - we now have 2 listeners for that notification. 
4. Repeat step 1 you have 4 listeners, then 8, 16, 32, etc, etc 

Eventually everything is clean-up (once the toast hides) but with an unlucky sequence of focus changes and because this leak grows exponential things get messy quickly. I have reached states  with 500000+ listeners easily which means game over.

**Comment by jrieken** [comment]
I have created https://github.com/Microsoft/vscode/pull/63136 to fix this.

**Comment by jrieken** [comment]
**Everyone** that is affected by this: Clearing your notifications should help to workaround this.

**Comment by filipesilva** [comment]
I was seeing this while editing a file for which I did not have the suggested extension, which triggered a popup that I didn't bother clearing.

**Comment by jrieken** [comment]
I have pushed https://github.com/Microsoft/vscode/commit/aa4ef3a42c2240b22863c64ba972c08fa9b224f3 to master which should fix this for next insiders

**Comment by fusengine** [comment]
Vscode is Nice this frozen completly my mac hahahah

![24edbb38-8459-47e4-b3fc-e44d8694ab6a](https://user-images.githubusercontent.com/7295655/48491024-ef8a0000-e826-11e8-8ac2-74103695c8fd.jpeg)

**Comment by jrieken** [comment]
@fusion809 not nice but also unrelated to this issue. please create a separate issue for this (cc @roblourens)

**Comment by sgyzetrov** [comment]
Make sense. With freshly installed insider ver, I never experienced hang since I don't have plugins that popup errors.

**Comment by fusengine** [comment]
Window😻

**Comment by fusion809** [comment]
@jrieken Am I right in saying you pinged me by accident? That you were really aiming for @fusengine?

**Comment by fusengine** [comment]
> Yeah, sorry about about...

why aim for me?

**Comment by christoxl** [comment]
I just leave it unattended for 5 or more minutes and it freezes completely. 

I'm using Linux Mint 18.3 and this wasn't happening last week.

[enabled_extensions.txt](https://github.com/Microsoft/vscode/files/2582803/enabled_extensions.txt)
[system_information.txt](https://github.com/Microsoft/vscode/files/2582804/system_information.txt)

**Comment by roblourens** [comment]
We have pushed a possible fix to Insiders. Would appreciate if everyone checks out our insiders build to verify that you don't see this issue there: https://code.visualstudio.com/insiders/

**Comment by fusengine** [comment]
Your insider not change problem

**Comment by fusengine** [comment]
sublime text is much more stable than vscode it's a shame on the part of windows. before publishing a version, try the machine

**Comment by jrieken** [comment]
@fusengine According to your first comment (https://github.com/Microsoft/vscode/issues/62970#issuecomment-438693227) you are *not* having the problem discussed in this issue. There is a separate issue about many `rg` processes so please make sure to comment on the right issue

**Comment by fusengine** [comment]
> @fusengine According to your first comment ([#62970 (comment)](https://github.com/Microsoft/vscode/issues/62970#issuecomment-438693227)) you are _not_ having the problem discussed in this issue. There is a separate issue about many `rg` processes so please make sure to comment on the right issue
start to publish stable software

**Comment by fusengine** [comment]
why hide my comment 😹

**Comment by guillaumevincent** [comment]
Freeze on fedora 29 also

My extensions:
```
esbenp.prettier-vscode
ms-python.python
robertohuertasm.vscode-icons
```

**Comment by mateuszgazdziak** [comment]
How to revert to previous version?

**Comment by fusengine** [comment]
> How to revert to previous version?

@mateuszgazdziak do you have a lot `process rg`?

**Comment by lorenz068** [comment]
Same here on Mac OSX Mojave, since the 1.29.0 update

**Comment by JCMais** [comment]
Was having same issue, latest insiders seems to have fixed it, thanks for working on that.

**Comment by bsendras** [comment]
VSCode v1.29.0. hangs on Mac OSX High Sierra 10.13.6.

**Comment by jrieken** [comment]
@bsendras Yeah, we already know that. We work full steam towards 1.29.1. In the meantime you can use insiders: http://code.visualstudio.com/insiders/ which already has the fix for this.

**Comment by mateuszgazdziak** [comment]
@fusengine I am worikng on Windows 10 :(

**Comment by YevhenKysilov** [comment]
#63214

**Comment by jrieken** [comment]
>  This problem was in the insider version too. Why they released the stable version with this issue?

If you see "this" in latest insiders that you don't have *this* issue but another issue. That's unhappy but we are here to help and fix. Please file a *new* issue, bonus points for providing version string, os, etc. Last, please following our perf guide: https://github.com/Microsoft/vscode/wiki/Performance-Issues

**Comment by fusengine** [comment]
When you fix that ?

**Comment by oppoi1** [comment]
Still having the same issues after your very recent fix. Especially problematic when scrolling or getting new input in the integrated terminal. Also, switching from 'files' to 'search' causes a good lag(and about 25% cpu usage for a few secs)

**Comment by jrieken** [comment]
@oppoi1 This is about a memory leak that happens with uncleared notification. Your comment doesn't match that, so please file a new issue. Bonus points for following our guidelines: https://github.com/Microsoft/vscode/wiki/Performance-Issues

**Comment by potomek** [comment]
Today this was happend when **I pasted** into JavaScript doc. **sentence with spaces**. I suppose the code validation system froze.

**Comment by jrieken** [comment]
This issue is about a very specific, fast growing, memory leak that happens when having notifications. We have released a fix for this issue in 1.29.1 and 1.30-insiders. Please make sure to be on those or later version. 

*Different* issues often show the *same* symptoms but the reverse if often not true. That means if you see a slowdown (symptom) it's very likely not this issue but another issue. Please try latest, please file new issues when an issue is closed, please follow this guide: https://github.com/Microsoft/vscode/wiki/Performance-Issues
